### PR TITLE
Adds Azure GPT 3.5 

### PR DESCRIPTION
**Pull request by nadheesh**
## Purpose
resolves https://github.com/wso2-enterprise/choreo/issues/22514, https://github.com/wso2-enterprise/choreo/issues/22515, https://github.com/wso2-enterprise/choreo/issues/22584

### CODE REVIEW

File: ballerina/llm.bal [hunk 1]
```diff
@@ -30,27 +33,34 @@ public type Gpt3ModelConfig readonly & record {|
...
     # Temperature value to be used for the completion. Default is `0.7`.
     decimal temperature = DEFAULT_TEMPERATURE;
 |};
 
-type PromptConstruct record {|
+# Prompt construct record
+#
+# + instruction - Instructions in the prompt
+# + query - Query to the prompt
+# + history - Execution history to the prompt
```
Comment (JayaniH) [inline ballerina/llm.bal:62]: we need to update the documentation here to be consistent with other record types

File: ballerina/toolkit.bal [hunk 2]
```diff
@@ -168,53 +176,60 @@ public isolated class HttpServiceToolKit {
 
     isolated function getTools() returns Tool[]|error => self.tools;
 
-    private isolated function get(HttpInput httpInput) returns json|error {
+    private isolated function get(HttpInput httpInput) returns HttpOutput|error {
         string path = check getPathWithParams(httpInput.path, httpInput?.pathParameters, httpInput?.queryParameters);
         log:printDebug(string `HTTP GET ${path} ${httpInput?.requestBody.toString()}`);
         http:Response getResult = check self.httpClient->get(path, headers = self.headers);
-        return getResult.getTextPayload(); // TODO improve http:Client error response handling
+        string payload = check getResult.getTextPayload();
```
Comment (JayaniH) [inline ballerina/toolkit.bal:183]: what if the payload is json?

File: ballerina/toolkit.bal [hunk 3]
```diff
@@ -168,53 +176,60 @@ public isolated class HttpServiceToolKit {
...
+        return {code: getResult.statusCode, payload};
```
Comment (JayaniH) [inline ballerina/toolkit.bal:184]: what would happen if we return the http response here?

File: ballerina/llm.bal (hunk 1 above)
Comment (nadheesh) [reply ballerina/llm.bal]: fixed

File: ballerina/toolkit.bal (hunk 2 above)
Comment (nadheesh) [reply ballerina/toolkit.bal:183]: We will get the text version of that json.

File: ballerina/toolkit.bal (hunk 3 above)
Comment (nadheesh) [reply ballerina/toolkit.bal:184]: According to my understanding, `http:Response`, itself does not include any information. It provides methods to retrieve the payload from the response received from the server. So the agent will have to know how to process the `http:Response`, which is a bad design. 
On the other hand, we need to provide only the required information to the agent as much as possible. So we don't need to send the response object to the Agent, which is an unnecessary complexity.

File: ballerina/toolkit.bal (hunk 2 above)
Comment (sameerajayasoma) [reply ballerina/toolkit.bal:183]: See: https://lib.ballerina.io/ballerina/http/2.9.0#Response-getTextPayload

Comment (nadheesh) [reply ballerina/toolkit.bal:183]: Thanks. it seems, it is safe to use `getTextPayload` because, right now we handle anything that is a text.

**Review by JayaniH — APPROVED** [review]
LGTM

**Review by malithjayasinghe — APPROVED** [review]

File: ballerina/llm.bal [hunk 4]
```diff
@@ -169,34 +198,92 @@ public isolated class ChatGptModel {
...
 
-    isolated function generate(PromptConstruct prompt) returns string|error {
-        chat:ChatCompletionRequestMessage[] messages = self.createPrompt(prompt);
-        return check self.chatComplete(messages, stop = OBSERVATION_KEY);
+    # Generate ReAct response for the given prompt
+    #
+    # + prompt - Prompt construct
+    # + return - ReAct response
+    public isolated function generate(PromptConstruct prompt) returns string|error {
+        return check self.chatComplete(createChatPrompt(prompt), stop = OBSERVATION_KEY);
```
Comment (MaryamZi) [inline ballerina/llm.bal:206]: `check` is unnecessary here and will add an unnecessary check.

File: ballerina/llm.bal [hunk 5]
```diff
@@ -169,34 +198,92 @@ public isolated class ChatGptModel {
...
+#
+# + prompt - Prompt construct
+# + return - ReAct prompt for chat LLMs
+public isolated function createChatPrompt(PromptConstruct prompt) returns ChatMessage[] {
+    string userMessage = "";
+    if prompt.history.length() == 0 {
+        userMessage = prompt.query;
+    }
+    else {
+        userMessage = string `${prompt.query}${"\n\n"}This was your previous work (but I haven\'t seen any of it! I only see what you return as final answer):${"\n"}`;
```
Comment (MaryamZi) [inline ballerina/llm.bal:275]:
```suggestion
    string userMessage;
    if prompt.history.length() == 0 {
        userMessage = prompt.query;
    } else {
        userMessage = string `${prompt.query}${"\n\n"}This was your previous work (but I haven\'t seen any of it! I only see what you return as final answer):${"\n"}`;
```

File: ballerina/llm.bal [hunk 6]
```diff
@@ -169,34 +198,92 @@ public isolated class ChatGptModel {
...
+        userMessage += constructHistoryPrompt(prompt.history);
```
Comment (MaryamZi) [inline ballerina/llm.bal:276]: Can't this just be a concatenation with the line above?

File: ballerina/toolkit.bal (hunk 2 above)
Comment (MaryamZi) [inline ballerina/toolkit.bal:183]: Shouldn't you check for status code and handle different codes/responses appropriately?

File: ballerina/toolkit.bal (hunk 3 above)
Comment (MaryamZi) [inline ballerina/toolkit.bal:184]: If you don't need the code can also directly use client data binding.

**Pull request merged**